{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "82e702e1",
   "metadata": {},
   "source": [
    "# Prepare raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eccfe90",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as ad\n",
    "\n",
    "# convenience functions for single cell analysis\n",
    "from sctools import plot, integrate, io, score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da67ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_status_txt = '''A3,inflamed\n",
    "A2,non-inflamed\n",
    "A1,healthy\n",
    "B3,inflamed\n",
    "B2,non-inflamed\n",
    "B1,healthy\n",
    "C3,inflamed\n",
    "C2,non-inflamed\n",
    "C1,healthy'''\n",
    "\n",
    "sample_status = {\n",
    "    s.split(',')[0]: s.split(',')[1]\n",
    "    for s\n",
    "    in sample_status_txt.split('\\n')\n",
    "}\n",
    "sample_status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3803b0ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\n",
    "    snakemake.input.matrix_tsv,\n",
    "    sep = '\\t',\n",
    "    compression = 'gzip',\n",
    "    index_col = 0\n",
    ")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5735de12",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "X = csr_matrix(data.T.values)\n",
    "X.eliminate_zeros()\n",
    "obs = pd.DataFrame(\n",
    "    index = data.columns\n",
    ")\n",
    "obs['sampleid'] = [i.split('-')[-1] for i in obs.index]\n",
    "obs['status'] = obs.sampleid.apply(\n",
    "    lambda x: sample_status[x]\n",
    ")\n",
    "var = pd.DataFrame(\n",
    "    index = data.index\n",
    ")\n",
    "adata = ad.AnnData(\n",
    "    X = X,\n",
    "    obs = obs,\n",
    "    var = var\n",
    ")\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f5f6060",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "source": [
    "# Filter genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbf1e5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "nexpressed_threshold = 10\n",
    "ngenes_passed = (adata.X.sum(axis = 0) >= nexpressed_threshold).sum()\n",
    "ngenes = adata.var.shape[0]\n",
    "print(\n",
    "    f'{ngenes_passed} of {ngenes} are retained requiring their expression in {nexpressed_threshold} or more cells'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89535cc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "filtered = adata[:, adata.X.sum(axis = 0) >= nexpressed_threshold].copy()\n",
    "filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e53a9f",
   "metadata": {},
   "source": [
    "# Plot raw data UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa3a70fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_raw_umap(adata):\n",
    "    tmp = adata.copy()\n",
    "    # don't need this here because it is already normalized and logged\n",
    "    # sc.pp.normalize_total(\n",
    "    #     tmp, \n",
    "    #     target_sum = 1e4\n",
    "    # )\n",
    "    # sc.pp.log1p(tmp)\n",
    "    sc.pp.pca(\n",
    "        tmp, \n",
    "        n_comps = 40, \n",
    "        svd_solver = 'arpack'\n",
    "    )\n",
    "    sc.pp.neighbors(\n",
    "        tmp,\n",
    "        use_rep = 'X_pca'\n",
    "    )\n",
    "    sc.tl.umap(tmp)\n",
    "    return tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a73529",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw = compute_raw_umap(filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "856cb9b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plot.misc.generate_and_save_for_figure(\n",
    "    raw,\n",
    "    {\n",
    "        'sampleid': (None, None), \n",
    "        'status': (snakemake.params.condition_palette, None),\n",
    "    },\n",
    "    snakemake.output.umap_raw,\n",
    "    snakemake.output.umap_raw_legend,\n",
    "    size = 40,\n",
    "    edgecolor = 'k',\n",
    "    linewidths = 0.3\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "752590bd",
   "metadata": {},
   "source": [
    "# Integrate with scVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "042f39ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "filtered.X = filtered.X.expm1()\n",
    "integrated = integrate.integrate_data_scvi(\n",
    "    filtered.copy(),\n",
    "    'sampleid',\n",
    "    train_size = 1\n",
    ")\n",
    "\n",
    "# integrated['data'].write(\n",
    "#     '../data/data.integrated.h5ad'\n",
    "# )\n",
    "\n",
    "# integrated['model'].save(\n",
    "#     '../data/data.integration.scvi.model',\n",
    "#     overwrite = True\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36781ffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "??plot.integrate.plot_integration_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b6f407",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, axs = plot.integrate.plot_integration_results(\n",
    "    {'data': integrated},\n",
    "    ['status', 'sampleid'],\n",
    "    [\n",
    "        dict(size = 10, vmax = None),\n",
    "        dict(size = 10, vmax = None)\n",
    "    ],\n",
    "    data_key='data'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5bf9b12",
   "metadata": {},
   "source": [
    "# Plot integrated UMAPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b86d2eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.misc.generate_and_save_for_figure(\n",
    "    integrated[\"data\"],\n",
    "    {\n",
    "        'sampleid': (None, None), \n",
    "        'status': (snakemake.params.condition_palette, None),\n",
    "    },\n",
    "    snakemake.output.umap_integrated,\n",
    "    snakemake.output.umap_integrated_legend,\n",
    "    size = 40,\n",
    "    edgecolor = 'k',\n",
    "    linewidths = 0.3,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87a0825",
   "metadata": {},
   "source": [
    "# Cell type annotation with celltypist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16bd51db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is the recommended resolution for majority vote in celltypist\n",
    "# we need to do this manually on integrated because integrated is only hvg\n",
    "resolution = 10\n",
    "sc.tl.leiden(\n",
    "    integrated[\"data\"],\n",
    "    key_added = f'leiden_scvi_{resolution}',\n",
    "    resolution = resolution\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4abb5545",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# reinitialize from raw to get all genes\n",
    "bdata = io.initialize_from_raw(integrated[\"data\"])\n",
    "\n",
    "# need log1p because saved data is expm1\n",
    "sc.pp.log1p(bdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7447901",
   "metadata": {},
   "outputs": [],
   "source": [
    "import celltypist as ct\n",
    "model = ct.models.Model.load(model=snakemake.params.celltypist_model)\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a59758a",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = ct.annotate(\n",
    "    bdata,\n",
    "    model=snakemake.params.celltypist_model,\n",
    "    majority_voting = True,  # following the celltypist tutorial\n",
    "    over_clustering = f'leiden_scvi_{resolution}'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "450d0944",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdata = predictions.to_adata()\n",
    "integrated[\"data\"].obs['cell_type_coarse'] = cdata.obs.majority_voting\n",
    "integrated[\"data\"].obs['cell_type_fine'] = cdata.obs.predicted_labels  # the label pre-majority voting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be4d03b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    integrated[\"data\"],\n",
    "    color = ['cell_type_coarse', 'cell_type_fine', 'LGR5'],\n",
    "    legend_loc = 'on data',\n",
    "    frameon = False,\n",
    "    size = 20,\n",
    "    edgecolor = 'white',\n",
    "    lw = 0.2,\n",
    "    vmax = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ddf5386",
   "metadata": {},
   "outputs": [],
   "source": [
    "(integrated[\"data\"].obs.cell_type_fine == 'Stem cells').sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d06531d9",
   "metadata": {},
   "source": [
    "# Plot annotation UMAPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45e068b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = integrated[\"data\"]\n",
    "celltype_color_palettes = {\n",
    "    k: v for k, v \n",
    "    in zip(\n",
    "        adata.obs.cell_type_coarse.cat.categories, \n",
    "        adata.uns['cell_type_coarse_colors']\n",
    "    )\n",
    "}\n",
    "celltype_color_palettes['Stem cells'] = '#EF5A7E'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93622fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "selected_cell_types = [\n",
    "    y if y == 'Stem cells' else x \n",
    "    for x, y \n",
    "    in zip(\n",
    "        adata.obs['cell_type_coarse'].values.to_list(), \n",
    "        adata.obs['cell_type_fine'].values.to_list()\n",
    "    )\n",
    "]\n",
    "\n",
    "adata.obs['cell_type_select'] = selected_cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97958855",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.misc.generate_and_save_for_figure(\n",
    "    adata,\n",
    "    {\n",
    "        'cell_type_coarse': (celltype_color_palettes, None),\n",
    "        'cell_type_select': (celltype_color_palettes, None),\n",
    "    },\n",
    "    snakemake.output.umap_annotated,\n",
    "    snakemake.output.umap_annotated_legend,\n",
    "    size = 40,\n",
    "    edgecolor = 'k',\n",
    "    linewidths = 0.3\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f24a8b0",
   "metadata": {},
   "source": [
    "# Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de6dea28",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = io.initialize_from_raw(\n",
    "    integrated[\"data\"]\n",
    ")\n",
    "sc.pp.log1p(adata)\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby = 'cell_type_coarse',\n",
    "    groups = ['TA'],\n",
    "    reference = 'rest',\n",
    "    method = 'wilcoxon',\n",
    "    corr_method = 'benjamini-hochberg'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fddba13",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "de_ta = pd.DataFrame.from_dict(\n",
    "    dict(\n",
    "        genes = [name[0] for name in adata.uns['rank_genes_groups']['names']],\n",
    "        padj = [p[0] for p in adata.uns['rank_genes_groups']['pvals_adj']],\n",
    "        lfc = [lfc[0] for lfc in adata.uns['rank_genes_groups']['logfoldchanges']]\n",
    "    )\n",
    ")\n",
    "de_ta['-log10padj'] = -np.log10(de_ta.padj)\n",
    "de_ta.index = de_ta.genes\n",
    "ax = sns.scatterplot(\n",
    "    data = de_ta,\n",
    "    x = 'lfc',\n",
    "    y = '-log10padj'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aa9025d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    adata,\n",
    "    groupby = 'cell_type_fine',\n",
    "    groups = ['Stem cells'],\n",
    "    reference = 'rest',\n",
    "    method = 'wilcoxon',\n",
    "    corr_method = 'benjamini-hochberg'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d15296d",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_sc = pd.DataFrame.from_dict(\n",
    "    dict(\n",
    "        genes = [name[0] for name in adata.uns['rank_genes_groups']['names']],\n",
    "        padj = [p[0] for p in adata.uns['rank_genes_groups']['pvals_adj']],\n",
    "        lfc = [lfc[0] for lfc in adata.uns['rank_genes_groups']['logfoldchanges']]\n",
    "    )\n",
    ")\n",
    "de_sc['-log10padj'] = -np.log10(de_sc.padj)\n",
    "de_sc.index = de_sc.genes\n",
    "ax = sns.scatterplot(\n",
    "    data = de_sc,\n",
    "    x = 'lfc',\n",
    "    y = '-log10padj'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11a7443d",
   "metadata": {},
   "source": [
    "# Compute stemness score as described by Tirosh et al. 2016"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22fa2500",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_index = (adata.obs.cell_type_select == 'Stem cells') & (adata.obs.status != 'healthy')\n",
    "stem_cells = adata[cell_index, :]\n",
    "stem_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b4a94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = set(stem_cells.var.index) & set(snakemake.params.stemness_markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d7bc9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "score_df = pd.DataFrame(\n",
    "    {\n",
    "        'stemness_score': score.gene_module_score(stem_cells, list(genes)),\n",
    "        'status': ['inflamed' if x == 'inflamed' else 'non-inflamed' for x in stem_cells.obs.status]\n",
    "    },\n",
    "    index = stem_cells.obs.index\n",
    ")\n",
    "score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f14ab39",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "tres = stats.ttest_ind(\n",
    "    score_df.loc[score_df.status == 'inflamed', 'stemness_score'],\n",
    "    score_df.loc[score_df.status == 'non-inflamed', 'stemness_score'],\n",
    "    alternative = 'less'\n",
    ")\n",
    "tres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3c478c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "sns.violinplot(\n",
    "    data = score_df,\n",
    "    y = 'status',\n",
    "    x = 'stemness_score',\n",
    "    ax = ax,\n",
    ")\n",
    "ax.text(\n",
    "    0.4,\n",
    "    0.5,\n",
    "    'p = {:.4e}'.format(tres.pvalue),\n",
    "    ha = 'right'\n",
    ")\n",
    "fig.set_figheight(3)\n",
    "fig.set_figwidth(6)\n",
    "fig.tight_layout()\n",
    "fig.savefig(snakemake.output.stemness_score_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12c27227",
   "metadata": {},
   "source": [
    "# Rank plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "885e25e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def map_group(x, lfc_cut, padj_cut):\n",
    "    signif =  (abs(x.lfc) > lfc_cut) & (x.padj < padj_cut)\n",
    "    \n",
    "    if signif:\n",
    "        return 'sig'\n",
    "    \n",
    "    return 'n.s.'\n",
    "\n",
    "\n",
    "def label_point(poi, ax):\n",
    "    x = poi['rank']\n",
    "    y = poi['lfc']\n",
    "    ax.plot(\n",
    "        [x, 600],\n",
    "        [y] * 2,\n",
    "        c = 'k',\n",
    "        lw = 0.5\n",
    "    )\n",
    "    gene_name = poi['genes']\n",
    "    padj = poi['padj']\n",
    "    \n",
    "    ax.text(\n",
    "        600, y,\n",
    "        f'{gene_name} padj = {padj:.3e}',\n",
    "        va = 'center',\n",
    "        ha = 'left'\n",
    "    )\n",
    "    \n",
    "\n",
    "palette = {\n",
    "    'sig': (178/256, 24/256, 43/256),\n",
    "    'n.s.': 'lightgrey'\n",
    "}\n",
    "\n",
    "zorders = {\n",
    "    'sig': 2,\n",
    "    'n.s.': 1\n",
    "}\n",
    "\n",
    "fig, axs = plt.subplots(2)\n",
    "\n",
    "for ax, df, title, label in zip(\n",
    "    axs,\n",
    "    [de_ta, de_sc],\n",
    "    ['TA cells', 'stem cells'],\n",
    "    ['TA', 'SC']\n",
    "):\n",
    "    df = df.loc[df.lfc > 0, :]\n",
    "    df['rank'] = df.lfc.rank(ascending = False)\n",
    "    df['-log10padj'] = -np.log10(df.padj)\n",
    "    df['group'] = df.apply(\n",
    "        map_group,\n",
    "        axis = 1,\n",
    "        lfc_cut = 1,\n",
    "        padj_cut = 1e-4\n",
    "    )\n",
    "    for g, gdf in df.groupby('group'):\n",
    "        ax.scatter(\n",
    "            x = gdf['rank'],\n",
    "            y = gdf['lfc'],\n",
    "            color = palette[g],\n",
    "            zorder = zorders[g],\n",
    "            label = g,\n",
    "            edgecolors = 'white',\n",
    "            linewidths = 0.5\n",
    "        )\n",
    "        \n",
    "        ax.axhline(\n",
    "            1,\n",
    "            ls = '--',\n",
    "            c = 'grey',\n",
    "            lw = 1\n",
    "        )\n",
    "    \n",
    "    label_point(df.loc['LGR5', :], ax)\n",
    "    \n",
    "    ax.legend()\n",
    "    ax.set_ylabel('log2FC')\n",
    "    ax.set_xlabel('log2FC rank')\n",
    "    ax.set_title(title)\n",
    "            \n",
    "fig.set_figwidth(3)\n",
    "fig.set_figheight(6)\n",
    "fig.tight_layout()\n",
    "fig.savefig(snakemake.output.rank_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa280ba8",
   "metadata": {},
   "source": [
    "# Cellwhisperer embedding UMAPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd253310",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "cwdata = sc.read_h5ad(snakemake.input.cellwhisperer_dataset)\n",
    "cwdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b692823",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwdata.obsm['X_umap'] = cwdata.obsm['X_cellwhisperer_umap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dc6ecab",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.misc.generate_and_save_for_figure(\n",
    "    cwdata,\n",
    "    {\n",
    "        'sample_id': (None, None), \n",
    "        'condition': (snakemake.params.condition_palette, None),\n",
    "    },\n",
    "    snakemake.output.umap_cellwhisperer,\n",
    "    snakemake.output.umap_cellwhisperer_legend,\n",
    "    size = 40,\n",
    "    edgecolor = 'k',\n",
    "    linewidths = 0.3\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanalysis",
   "language": "python",
   "name": "scanalysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
